### PR TITLE
METRON-1834: Migrate Elasticsearch from TransportClient to new Java REST API

### PR DESCRIPTION
**Pull request by mmiklavc**
## Contributor Comments

https://issues.apache.org/jira/browse/METRON-1834

This task has been a long time coming after having completed the ES upgrade in https://issues.apache.org/jira/browse/METRON-939. Motivation for completing this now is that Elasticsearch will be deprecating use of the TransportClient in v 7.x. This PR migrates the Elasticsearch client from TransportClient to the newer Java REST API. 

1. https://www.elastic.co/guide/en/elasticsearch/client/java-api/5.6/client.html
2. https://www.elastic.co/guide/en/elasticsearch/client/java-rest/5.6/java-rest-overview.html
3. https://www.elastic.co/guide/en/elasticsearch/client/java-rest/5.6/java-rest-high-level-migration.html

This builds off and finishes work started by @cestella here - https://github.com/cestella/incubator-metron/tree/es_rest_client. I condensed his branch into 1 flattened commit and built on top of it in order to provide attribution.

I have a number of tasks I'm still working through, but I wanted to get the review process started. I've minimally validated X-Pack auth and will have some follow-up for SSL. Test plans and a breakdown of the changes will be soon to follow. For starters, full dev should continue to work as normal and you should see data flowing into indexes for bro, snort, and yaf. There are some additional changes to how this client will be configured, which I'll be documenting shortly. The new client does not take a Map of settings any longer now that it is leverage Apache HTTP Async Client https://www.elastic.co/guide/en/elasticsearch/client/java-rest/5.6/java-rest-low-usage-dependencies.html under the hood. This meant choosing a set of properties to expose and doing a translation to the builder pattern under the hood. Again, I'll have a write-up of this in the migration guide and update the README's accordingly.

NOTE: This checks off 2 items from this follow-on list https://github.com/apache/metron/pull/840#issuecomment-347281776

1. Fix Log4j logging problem - classpath issues
2. Migrate to new ES REST client

Per discussion in the Metron Slack channel, I will be updating the Jira ticket with a series of tasks to be completed prior to acceptance, including performance regression testing compared with the old API.

## Pull Request Checklist

Thank you for submitting a contribution to Apache Metron.  
Please refer to our [Development Guidelines](https://cwiki.apache.org/confluence/pages/viewpage.action?pageId=61332235) for the complete guide to follow for contributions.  
Please refer also to our [Build Verification Guidelines](https://cwiki.apache.org/confluence/display/METRON/Verifying+Builds?show-miniview) for complete smoke testing guides.  

In order to streamline the review of the contribution we ask you follow these guidelines and ask you to double check the following:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? If not one needs to be created at [Metron Jira](https://issues.apache.org/jira/browse/METRON/?selectedTab=com.atlassian.jira.jira-projects-plugin:summary-panel).
- [x] Does your PR title start with METRON-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.
- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

### For code changes:
- [x] Have you included steps to reproduce the behavior or problem that is being changed or addressed?
- [x] Have you included steps or a guide to how the change may be verified and tested manually?
- [x] Have you ensured that the full suite of tests and checks have been executed in the root metron folder via:
  ```
  mvn -q clean integration-test install && dev-utilities/build-utils/verify_licenses.sh 
  ```

- [x] Have you written or updated unit tests and or integration tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [x] Have you verified the basic functionality of the build by building and running locally with Vagrant full-dev environment or the equivalent?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered by building and verifying the site-book? If not then run the following commands and the verify changes via `site-book/target/site/index.html`:

  ```
  cd site-book
  mvn site
  ```

#### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.
It is also recommended that [travis-ci](https://travis-ci.org) is set up for your personal repository such that your branches are built there before submitting a pull request.

### CODE REVIEW

**Comment by mmiklavc** [comment]
A recent merge with https://github.com/apache/metron/pull/1218 has caused a number of integration test failures. Looking into it.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/FieldProperties.java [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+ * limitations under the License.
+ */
+package org.apache.metron.elasticsearch.utils;
+
+import org.apache.commons.collections4.map.AbstractMapDecorator;
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class FieldProperties extends AbstractMapDecorator<String, Object> {
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/FieldProperties.java]: Can you add javadoc on the class?

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java [hunk 2]
```diff
@@ -76,7 +76,8 @@ public BulkWriterResponse write(String sensorType, WriterConfiguration configura
     FieldNameConverter fieldNameConverter = FieldNameConverters.create(sensorType, configurations);
 
     final String indexPostfix = dateFormat.format(new Date());
-    BulkRequestBuilder bulkRequest = client.prepareBulk();
+    BulkRequest bulkRequest = new BulkRequest();
+    //BulkRequestBuilder bulkRequest = client.prepareBulk();
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java]: Can we remove this commented out line?

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+import org.apache.metron.elasticsearch.utils.FieldMapping;
+import org.apache.metron.elasticsearch.utils.FieldProperties;
+import org.elasticsearch.client.Response;
+import org.elasticsearch.client.RestClient;
+import org.elasticsearch.client.RestHighLevelClient;
+
+/**
+ * Wrapper around the Elasticsearch REST clients. Exposes capabilities of the low and high-level clients.
+ */
+public class ElasticsearchClient implements AutoCloseable{
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: Any wisdom or linkage to help developers understand when they should use the high-level versus low-level client?

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            , entity
+    );
+
+    if(response.getStatusLine().getStatusCode() != 200) {
+      String responseStr = IOUtils.toString(response.getEntity().getContent());
+      throw new IllegalStateException("Got a " + response.getStatusLine().getStatusCode() + " due to " + responseStr);
+    }
+  }
+
+  public String[] getIndices() throws IOException {
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: Are there unit tests specifically for this class?  I don't see any, but maybe I missed.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java (hunk 3 above)
Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: Actually, I don't see any uses of `ElasticsearchClient.getLowLevelClient()`.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchColumnMetadataDao.java [hunk 5]
```diff
@@ -64,12 +62,12 @@
   /**
    * An Elasticsearch administrative client.
    */
-  private transient AdminClient adminClient;
+  private transient ElasticsearchClient adminClient;
```
[the remark this answers is not in the document]
Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/dao/ElasticsearchColumnMetadataDao.java]: Same in `ElasticsearchRetrieveLatestDao.transportClient`.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java [hunk 6]
```diff
@@ -124,102 +129,146 @@ public static String getBaseIndexName(String indexName) {
...
-      for (HostnamePort hp : getIps(globalConfiguration)) {
-        client.addTransportAddress(
-                new InetSocketTransportAddress(InetAddress.getByName(hp.hostname), hp.port)
-        );
-      }
-      return client;
-    } catch (UnknownHostException exception) {
-      throw new RuntimeException(exception);
-    }
+  public static ElasticsearchClient getClient(Map<String, Object> globalConfiguration) {
```
Comment (nickwallen) [inline metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java]: I just found myself trying to instantiate a new `ElasticsearchClient` with no idea how to go about doing that.  Then I remember running across this method in your PR.  This seems really broadly useful.

Could we make this functionality more discoverable somehow?  We could make this a static method on `ElasticsearchClient`.  Or just refer to this method in the javadocs for `ElasticsearchClient`?  Or some other way?

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  @Override
+  public void close() throws IOException {
+    if(lowLevelClient != null) {
+      lowLevelClient.close();
+    }
+  }
+
+  public void putMapping(String index, String type, String source) throws IOException {
```
[the remark this answers is not in the document]
Comment (nickwallen) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: For example, I was just trying to use this.  I assume type == docType, but that's not completely clear to me.

**Comment by mmiklavc** [comment]
I'm happy to clarify variable and method names here, but I want to be careful about overdoing it on the javadoc. It gets stale, is not testable, and is far more often overlooked than the code itself. I'm all for javadoc on public API's.

**Comment by cestella** [comment]
Just chiming in here in support of taking some care around javadoc.  I tend to agree that it goes stale and it can become misleading.  I think the metric (and this is just personal opinion here) should be that any public API should be well javadoc'd and otherwise documented (e.g. the REST API) and we should seriously consider javadoc'ing classes which are used broadly.

I tend to prefer a compromise whereby the class itself has documentation around the intent and use of the class in Javadoc and the methods are named well and broken up well.  From there, I'm ok if there's less effort put around javadoc of individual methods.

Just my $0.02

**Comment by mmiklavc** [comment]
Thanks for the feedback/viewpoint @cestella . I think @nickwallen has some good points, in particular about why/when to choose one client or the other. I'm adding some of that detail now, and am linking aggressively to the ES docs so that we're not simply reproducing their documentation. In general, I do NOT think we should be writing javadoc like this - https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html. It is certainly detailed, but arguably pedantic for our purposes. Oh, and we should also be documenting Stellar functions - which is a feature we already support via the annotations you added some time ago.

**Comment by nickwallen** [comment]
I get your high-level take on DRY.  Sure, makes sense.  And I'm certainly not asking for anything close to the level of detail in the Arrays example.  

I just so often find myself tracing through code for 15 minutes to find answers to questions that could have been answered with one, well-placed comment. Spent most of the day in that situation actually.  That being said, whatever middle ground you can find will likely suffice.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/config/ElasticsearchClientOptions.java [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.metron.elasticsearch.config;
+
+import org.apache.metron.common.configuration.ConfigOption;
+
+public enum ElasticsearchClientOptions implements ConfigOption {
```
[the remark this answers is not in the document]
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/config/ElasticsearchClientOptions.java]: One is an enum of the options for the elasticsearch client. The other is the config used to access the configuration settings provided via global config (or other method, if a different context were ever used).

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java (hunk 6 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java]: I'll refactor this a bit.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java (hunk 3 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: low level client is for exposing externally, if necessary. The current uses are internal to the class as a member variable.

**Comment by mmiklavc** [comment]
I've started addressing some of the questions you had @nickwallen. Not quite finished yet, but the rest should be soon to follow. I'd like to clean up the way we handle the ElasticsearchClient and its instantiation a bit.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java (hunk 6 above)
Comment (ottobackwards) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java]: Maybe this should be a builder pattern

**Comment by mmiklavc** [comment]
I think we're all on the same page @ottobackwards, @nickwallen, and @cestella . As I mentioned early on, this effort was somewhat unique as it was built off of some POC work from @cestella from a couple months back, and I wanted to enhance that work (e.g. addressing the TODO's, bugs, and configuration issues mentioned) while refactoring/changing as little as possible. That's on me for not having gotten out in front of polishing a few of the common/reusable pieces a bit more, so your comments are well received. Updates soon to follow.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java (hunk 6 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/utils/ElasticsearchUtils.java]: @ottobackwards I was looking at that, but with the callbacks in the ES client API, a builder pattern is really difficult to use here.

**Comment by mmiklavc** [comment]
@ottobackwards @nickwallen - just to shed some additional light on the choices made here, take a look at the DAO classes. The new `ElasticsearchClient` Metron client class lays some foundation towards wrapping the ES API. What this PR explicitly does *not* do is completely abstract away the ES dependencies. I believe that's a task that we should perform, but I do not believe it makes sense to couple it with the task of swapping out the client.

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java (hunk 2 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/writer/ElasticsearchWriter.java]: Done

**Comment by mmiklavc** [comment]
@nickwallen I updated the ElasticsearchSearchIntegrationTest a bit. I'm now using the new client to setup the bro/snort templates, create the indices, and load the data.

**Comment by nickwallen** [comment]
@mmiklavc Keep in mind that I already have the open PR that refactors the data load in those tests to use the `ElasticsearchUpdateDao` #1247.   I will work through the conflicts after this goes in.

Are you happy with this now?  Is it done done?

**Comment by mmiklavc** [comment]
## Testing Notes For X-Pack AND SSL (I know, gettin fancy):

1. Install X-Pack

    This will also install the certgen tool which can be used for generating certificates for SSL.

    ```
    /usr/share/elasticsearch/bin/elasticsearch-plugin install x-pack
    ```

1. Setup X-Pack User

    username = xpack_client_user 
    password = changeme 
    role = superuser

    ```
    sudo /usr/share/elasticsearch/bin/x-pack/users useradd xpack_client_user -p changeme -r superuser
    ```

1. Setup cert using the ES certgen tools

    Run the certgen tool. You can store the certs in /etc/elasticsearch/ssl/certs.zip when prompted. Use "node1" as the instance name in fulldev because you'll want it to match your host. You can leave the IP and DNS details blank.

    ```
    /usr/share/elasticsearch/bin/x-pack/certgen --pass
    ```

    Extract the certs

    ```
    cd /etc/elasticsearch/ssl
    unzip certs.zip
    # I flattened all ca/certs so it looks as follows
    ls -1 /etc/elasticsearch/ssl
    ca.crt
    ca.key
    esnode.crt
    esnode.key
    ```

1. Setup Elasticsearch to use the certs

    https://www.elastic.co/guide/en/x-pack/5.6/ssl-tls.html

    1. Modify /etc/elasticsearch/elasticsearch.yml

        ```
        xpack.ssl.key:                     /etc/elasticsearch/ssl/esnode.key
        xpack.ssl.certificate:             /etc/elasticsearch/ssl/esnode.crt
        xpack.ssl.certificate_authorities: [ "/etc/elasticsearch/ssl/ca.crt" ]
        xpack.security.transport.ssl.enabled: true
        xpack.security.http.ssl.enabled: true
        ```

1. Setup the client truststore

    1. Import the Certificate Authority (CA). 

        * Specify an alias of your choosing. I chose "elasticCA". 
        * You'll also be prompted for a password, which must be at least 6 characters. I used "apachemetron".
        * When prompted to "Trust this certificate?" type yes and hit enter.

            ```
            keytool -import -alias elasticCA -file /etc/elasticsearch/ssl/ca.crt -keystore clienttruststore.jks
            ```

    1. Import the data node certificate.
        
        * Specify an alias of your choosing. In fulldev use "node1" as it will need to match your hostname.
        * Enter the password that you set when creating the truststore.

            ```
            keytool -importcert -keystore clienttruststore.jks -alias node1 -file /etc/elasticsearch/ssl/esnode.crt
            ```

    1. Put the truststore in the storm user home dir for the purpose of our tests.

        ```
        mv clienttruststore.jks /home/storm/
        chown storm:hadoop /home/storm/clienttruststore.jks
        ```
    
1. Configure the Elasticsearch client in Metron

    1. Load the passwords in HDFS for x-pack and the truststore

        ```
        echo changeme > /tmp/xpack-password
        echo apachemetron > /tmp/truststore-password
        sudo -u hdfs hdfs dfs -mkdir /apps/metron/elasticsearch/
        sudo -u hdfs hdfs dfs -put /tmp/xpack-password /apps/metron/elasticsearch/
        sudo -u hdfs hdfs dfs -put /tmp/truststore-password /apps/metron/elasticsearch/
        sudo -u hdfs hdfs dfs -chown metron:metron /apps/metron/elasticsearch/*
        ```
    
    1. Modify the Metron global config with the SSL and X-Pack properties

        * Pull down the latest global config

            ```
            $METRON_HOME/bin/zk_load_configs.sh -m PULL -o $METRON_HOME/config/zookeeper -z $ZOOKEEPER -f

            ```

        * Update the configuration by adding the es.client.settings for xpack and SSL.

            ```
            "es.client.settings" : {
                "xpack.username" : "xpack_client_user",
                "xpack.password.file" : "/apps/metron/elasticsearch/xpack-password",
                "ssl.enabled" : true,
                "keystore.type" : "jks",
                "keystore.path" : "/home/storm/clienttruststore.jks",
                "keystore.password.file" : "/apps/metron/elasticsearch/truststore-password"
            }
            ```

        * Push the changes to Zookeeper

            ```
            $METRON_HOME/bin/zk_load_configs.sh -m PUSH -i $METRON_HOME/config/zookeeper -z $ZOOKEEPER
            # Confirm your changes are there
            $METRON_HOME/bin/zk_load_configs.sh -z $ZOOKEEPER -m DUMP -c GLOBAL
            ```

1. Restart Elasticsearch

    __Note:__ The custom properties were never, and still are not exposed via the Elasticsearch MPack. Restarting via Ambari will overwrite any changes you make manually.

    `service elasticsearch start`

    * You should be able to connect to Elasticsearch via SSL in your browser: `https://node1:9200/_cat/health?v`.
    * The Head plugin should also allow you to connect by slightly changing your URL to use https for SSL: `https://node1:9200/`

1. Get rid of existing indexes so that we can easily tell that new data is flowing into Elasticsearch.


    __Note:__ Use "node1" for the hostname. Otherwise, the cert will fail, triumphantly announcing `curl: (51) SSL: certificate subject name 'node1' does not match target host name 'localhost'`.

    ```
    curl --cacert /etc/elasticsearch/ssl/ca.crt --user xpack_client_user:changeme -XDELETE "https://node1:9200/squid*"
    curl --cacert /etc/elasticsearch/ssl/ca.crt --user xpack_client_user:changeme -XDELETE "https://node1:9200/bro*"
    curl --cacert /etc/elasticsearch/ssl/ca.crt --user xpack_client_user:changeme -XDELETE "https://node1:9200/snort*"
    curl --cacert /etc/elasticsearch/ssl/ca.crt --user xpack_client_user:changeme -XDELETE "https://node1:9200/yaf*"
    ```

1. Restart the indexing topology

    Ambari > Metron > Summary > Metron Indexing, choose restart

File: metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java (hunk 4 above)
Comment (mmiklavc) [reply metron-platform/metron-elasticsearch/src/main/java/org/apache/metron/elasticsearch/client/ElasticsearchClient.java]: Not unit tests, but the integration tests definitely cover this bit.

**Comment by nickwallen** [comment]
+1 Big effort and many thanks for this.  I ran this up in Full Dev and it worked like a charm.

**Comment by nickwallen** [comment]
@mmiklavc +1 remains. Addition to upgrade doc looks good.  A user may just have to translate from old properties to new.

**Comment by anandsubbu** [comment]
> The last remaining bit before I merge this will be sharing some results around the performance/regression testing performed by @anandsubbu.

I completed a set of performance comparison tests between the REST and Transport clients. Here's the update.

### Note on the Testing Done
* The performance regression comparison tests are run on a multi-node VM-based environment by:
a) Running a set of tests using the ES REST Client 
b) Swap out the REST client with the Transport client, and re-run the same set of tests.

### Observations
1) The indexing rate numbers with the REST client is similar to that of the Transport client.
2) I had noticed storm rebalance error messages for the kafkaSpout for both REST and Transport client runs, which required to adjust the values of `request.timeout.ms` and `session.timeout.ms` in the ES flux file under 'kafkaProps'. The errors were resolved after making these changes. 
3) For the REST client, I observed listener timeout exceptions with the default settings. I had to increase max retry and socket timeout values in global config as follows, in order to resolve the exceptions.
```
  "es.client.settings" : {
    "max.retry.timeout.millis" : "60000",
    "socket.timeout.millis" : "60000"
  }
```
4) For both cases, the kafkaSpout complete latency was seen to be on the higher side (~30 secs)
5) Lastly, I noticed that there were a number of failed tuples seen in the kafkaSpout of the RA Indexing topology. I am continuing to dig more into this, but since this was seen for both the REST and Transport client tests, I am suspecting it could be an anomaly with the test environment, or a different issue common to both REST/Transport clients.

Based on the above, it is seen that there is no immediate regression in performance b/w the REST and Transport client. 

+1, great work @mmiklavc

**Comment by mmiklavc** [comment]
I'm about to merge this and wanted to give some added feedback. @anandsubbu and I made some additional changes to the tuning options, specifically setting `topology.ackers.executors` equal to the number of inbound Kafka topic partitions, and we were able to see runs with no failures.

**Comment by nickwallen** [comment]
Ship it!

**Comment by anandsubbu** [comment]
> specifically setting `topology.ackers.executors` equal to the number of inbound Kafka topic partitions

And the number of Workers too. i.e. 
No. of Kafka partitions = No. of workers (`ra_indexing_workers`) = No. of Ackers (`ra_indexing_acker_executors`)

In my test, I set this to 10 and saw no failed tuples in the kafkaSpout.

My +1 holds. Thanks @mmiklavc !

**Pull request closed without merging**